### PR TITLE
Support relative path for snippet's filename attribute

### LINKED ISSUE #320: Support relative paths for snippet's filename attribute - only absolute works now
### Discussed in https://github.com/knqyf263/pet/discussions/317

<div type='discussions-op-text'>

<sup>Originally posted by **csessh** October  4, 2024</sup>
Let's say I have this config:
 
```
[General]
Backend = "gist"
Cmd = []
Color = true
Column = 50
Editor = "nvim"
Format = "[$tags] [$description] > $command"
SelectCmd = "fzf --ansi --highlight-line --layout=reverse --border --height=90% --pointer=* --cycle --prompt=Snippets:"
SnippetDirs = []
SnippetFile = "~/.config/pet/snippet.toml"
SortBy = "recency"
```

I specify a snippet.toml file path like so:
```
SnippetFile = "~/.config/pet/snippet.toml"
```

`pet new` would create an entry like this: 
```
[[Snippets]]
  Description = "Prune git repo"
  Filename = "/home/csessh/.config/pet/snippet.toml"
  Output = ""
  Tag = ["git"]
  command = "git fetch --all --prune"
```

Filename attribute is displayed as absolute path. This doesn't play too well with porting configs/snippets from one workstation to another if they don't share the same username.

Is there a way to enforce snippet entries to keep the relative format? 

or is there a particular reason for them to be saved in absolute path format? If there isn't, I'd be happy to submit a PR. </div>

### PR DESCRIPTION
**Pull request by csessh**
closes #320 

This pull request contains changes to support both relative and absolute path types for snippet file's filename attribute. 

Given the test case listed in #320: 

The same case presented in the issue

Given this config:

```toml
[General]
...
SnippetDirs = []
SnippetFile = "~/.config/test/test.toml"
...
```

`SnippetFile` is a relative path which expands to `/home/csessh/.config/test/test.toml`

Current Pet version:
```bash
fw13 :: ~/Documents/pet ‹feat-syspath*› » pet new 
Command> ping google.com
Description> hey Google
fw13 :: ~/Documents/pet ‹feat-syspath*› » cat ~/.config/test/test.toml 
   1   │ 
   2   │ [[Snippets]]
   3   │   Description = "hey Google"
   4   │   Filename = "/home/csessh/.config/test/test.toml"
   5   │   Output = ""
   6   │   Tag = []
   7   │   command = "ping google.com"
```

As expected, `~/.config/test/test.toml` expands to `/home/csessh/.config/test/test.toml`

Using feature branch version: 
```bash
fw13 :: ~/Documents/pet ‹feat-syspath*› » go run main.go new      
Command> ping google.com
Description> Hey Goolge
fw13 :: ~/Documents/pet ‹feat-syspath*› » go run main.go edit
fw13 :: ~/Documents/pet ‹feat-syspath*› » cat ~/.config/test/test.toml
   1   │ 
   2   │ [[Snippets]]
   3   │   Description = "Hey Goolge"
   4   │   Filename = "~/.config/test/test.toml"
   5   │   Output = ""
   6   │   Tag = []
   7   │   command = "ping google.com"
```

file path stays the same. The same result applies for the following test case:

```toml
[General]
...
SnippetDirs = []
SnippetFile = "/home/csessh/.config/test/test.toml"
...
```

Full path is rendered for every snippets.

I didn't go too deep into testing filename with the use of multiple directories as I ran into #321 problem.  However, I did include the safeguard check for it in this PR too. 

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

**Comment by csessh** [comment]
@RamiAwar could you review this PR, please?

File: cmd/new.go [hunk 1]
```diff
@@ -167,10 +167,15 @@ func createAndEditSnippet(newSnippet snippet.SnippetInfo, snippets snippet.Snipp
 
 func countSnippetLines() int {
 	// Count lines in snippet file
-	f, err := os.Open(config.Conf.General.SnippetFile)
+	if config.Conf.General.SnippetFile == "" {
```
Comment (RamiAwar) [inline cmd/new.go:170]: It might be nicer to handle the empty string case inside ExpandPath itself. 

No validation is taking place there, so right now the function ExpandPath is kind of buggy. Adding this empty string check there would make it more robust, and also make this code a bit nicer to read here.

File: config/config.go [hunk 2]
```diff
@@ -98,9 +98,8 @@ func (cfg *Config) Load(file string) error {
 			return err
 		}
 		var snippetdirs []string
-		cfg.General.SnippetFile = expandPath(cfg.General.SnippetFile)
```
Comment (RamiAwar) [inline config/config.go:101]: not directly clear to me why we need to remove the expand path from here

**Comment by RamiAwar** [comment]
How are you handling interaction with snippet directories?

File: cmd/new.go (hunk 1 above)
Comment (csessh) [reply cmd/new.go:170]: That's valid. Let me give it a go.

File: config/config.go (hunk 2 above)
Comment (csessh) [reply config/config.go:101]: This is where this issue is created. This line expand the path and this attribute is what gets written to file.

In this PR, my approach is to keep `cfg.General.SnippetFile` as is without expanding and expand whenever this file is to be checked/read/opened.

**Comment by csessh** [comment]
Directories can also be expressed in relative path format, e.g. `~/.config/pet/`. 
This PR doesn't touch this feature's logic, apart from expanding paths to ensure snippet files can be read/opened.

File: cmd/new.go [hunk 3]
```diff
@@ -166,11 +166,16 @@ func createAndEditSnippet(newSnippet snippet.SnippetInfo, snippets snippet.Snipp
 }
 
 func countSnippetLines() int {
-	// Count lines in snippet file
-	f, err := os.Open(config.Conf.General.SnippetFile)
+	filepath, err := config.ExpandPath(config.Conf.General.SnippetFile)
+	if err != nil {
+		return 0
+	}
```
Comment (csessh) [inline cmd/new.go:172]: As mentioned in #321, I made an assumption that if SnippetFile is omitted, it should simply return 0 for line count instead of panicking. 

Perhaps whoever understands multiple directories better could correct me here.

File: config/config.go (hunk 2 above)
Comment (RamiAwar) [reply config/config.go:101]: Ah I thought so. Makes sense

File: cmd/new.go (hunk 3 above)
Comment (RamiAwar) [reply cmd/new.go:172]: hmmm yes let me look into this further. 

I'm concerned about having a relative snippet file with directories, not sure how that would be handled.
That's even a concern in the current version of pet where you might have an absolute snippet file path as well as directories. Might not be an issue, just have to look into it and possibly document it.

Comment (csessh) [reply cmd/new.go:172]: I think having both `SnippetDirs` and `SnippetFile` configurable can be rather confusing due to the lack of signs to indicate which one of those two options should take priority.

I'd propose something like this:

```
SnippetFiles = ["~/.config/pet/snippet-1.toml", "/var/tmp/pet/snippet-2.toml" ]
```

This consolidates the option of managing multiple directories under one single configurable item. 
If you want to keep everything under a single file, specify a single snippet file.
If you want to manage multiple files, in various directories, specify multiple paths. 
When save/edit, a prompt to choose which file to save/edit is simple enough. 

Of course, I image it'd be a breaking change.

Comment (csessh) [reply cmd/new.go:172]: @RamiAwar I could take a jab at this approach later this week if you think it's worth considering

Comment (RamiAwar) [reply cmd/new.go:172]: @csessh sorry I still didn't look into this more, I was actually typing up a reply at some point but lost the tab 😓 

I think it's not a problem and there's no work necessary on your end, but I want to better document this. Will respond in a few days time!

Comment (RamiAwar) [reply cmd/new.go:172]: Going to add some more context here - will add it to the docs later as well. You don't have to act on this, just FYI, going through this to review your PR properly.

Having the main snippet file is mandatory.

If snippet directories are also configured, then the snippet Load method will load snippets from the main file + all files inside the snippet directory!

**What does this mean for:**
**1- Edits:** How do we edit a snippet if it can be part of any of several files? -> We make the user select the snippet they want to edit first, which tells us which file we should open for editing.

This is still not amazing UX because we send them to the beginning of the file and not the correct snippet line - this should be fixed. (should add a test for this, making sure we jump to the right line + fix it)

2- Search: We just search across all snippets from all files.

More fundamental operations:

3- Load: Loads snippets from the core file (mandatory) + all the files in all directories specified into one big list. Nice, expected.

4- Save: Given a slice of SnippetInfo, this loops over every snippet and checks its filename. This logic doesn't look right to me - we're just overwriting the snippetFile as we loop through? If it's empty, we're setting it to the FIRST snippet directory + a custom file name (???). 

I think here it was meant to be `snippet.Filename` maybe. This definitely seems like a bug, consistent with what you reported in that issue #321 I think.

image

Comment (RamiAwar) [reply cmd/new.go:172]: Do you think it makes sense for the sync to only sync the main snippet file and ignore files in the directories?

I want to try and avoid having to sync multiple files and keep track of each. Directories are nice, but I want to keep it a 'secondary' feature until I iron out the kinks. There are quite some bugs

Comment (RamiAwar) [reply cmd/new.go:172]: #323 fixes some of these

Comment (csessh) [reply cmd/new.go:172]: It'll be fine if it's communicated clearly in the tool's documentation that only main snippet file is sync'd. 
I, personally, don't use directories. sync or not.

**Comment by csessh** [comment]
Resolved conflicts

**Comment by RamiAwar** [comment]
Awesome, thank you! @csessh Will review this carefully this week. Thanks for waiting 🙏🏼 ✨

File: cmd/new.go (hunk 3 above)
Comment (RamiAwar) [reply cmd/new.go:172]: > As mentioned in #321, I made an assumption that if SnippetFile is omitted, it should simply return 0 for line count instead of panicking.
> 
> Perhaps whoever understands multiple directories better could correct me here.

Back to this - I think this needs a rebase now. New code actually panics here, which is what we want. Snippet file should always be present

File: config/config.go [hunk 4]
```diff
@@ -98,9 +99,12 @@ func (cfg *Config) Load(file string) error {
...
 		for _, dir := range cfg.General.SnippetDirs {
-			snippetdirs = append(snippetdirs, expandPath(dir)) // note the = instead of :=
+			if !strings.HasSuffix(dir, "/") {
```
Comment (RamiAwar) [inline config/config.go:103]: Why is this necessary?

File: config/config.go [hunk 5]
```diff
@@ -165,15 +177,25 @@ func GetDefaultConfigDir() (dir string, err error) {
...
+// Given a path to either a file or directory, returns its absolute path format.
+// ExpandPath resolves "~/" prefix in a given system path.
+// Raise error if path is an empty string as it
+func ExpandPath(path string) (string, error) {
+	if path == "" {
+		error := errors.New("path to file/directory is not set.")
+		return path, error
+	}
+
+	if strings.HasPrefix(path, "~/") {
```
Comment (RamiAwar) [inline config/config.go:189]: How would this be on Windows? Doesn't that use backslashes?

Maybe we should just check for tilde prefix instead and no slash, wdyt?

File: config/config.go [hunk 6]
```diff
@@ -165,15 +177,25 @@ func GetDefaultConfigDir() (dir string, err error) {
...
+		homedir, err := os.UserHomeDir()
+		if err != nil {
+			return path, err
 		}
+
+		return filepath.Join(homedir, path[2:]), nil
```
Comment (RamiAwar) [inline config/config.go:195]: I'd at least use a variable name for readability. Makes it even easier to see why path[:2]

`relativePath := path[:2]`

File: snippet/snippet.go [hunk 7]
```diff
@@ -30,45 +30,53 @@ func (snippets *Snippets) Load(includeDirs bool) error {
...
 
-	// Load snippets from the main snippet file
-	// Raise an error if the file is not found / not configured
-	snippetFile := config.Conf.General.SnippetFile
-	if snippetFile != "" {
+	snippetFile, err := config.ExpandPath(config.Conf.General.SnippetFile)
+	if err == nil {
 		if _, err := os.Stat(snippetFile); err == nil {
-			snippetFiles = append(snippetFiles, snippetFile)
+			snippetFiles = append(snippetFiles, config.Conf.General.SnippetFile)
```
Comment (RamiAwar) [inline snippet/snippet.go:36]: Why do you append the un-expanded path here instead of the expanded one? Not clear to me

File: snippet/snippet.go [hunk 8]
```diff
@@ -30,45 +30,53 @@ func (snippets *Snippets) Load(includeDirs bool) error {
...
+				return fmt.Errorf("snippet directory not found. %s", snippetDir)
+			}
+
 			if _, err := os.Stat(dir); err != nil {
 				if os.IsNotExist(err) {
-					return fmt.Errorf("snippet directory not found. %s", dir)
+					return fmt.Errorf("snippet directory not found. %s", snippetDir)
 				}
-				return fmt.Errorf("failed to load snippet directory. %v", err)
+				snippetFiles = append(snippetFiles, getFiles(dir)...)
```
Comment (RamiAwar) [inline snippet/snippet.go:60]: Why add this here? It is triggered right below

**Review by RamiAwar — CHANGES_REQUESTED** [review]

File: config/config.go (hunk 6 above)
Comment (csessh) [reply config/config.go:195]: That's reasonable

File: snippet/snippet.go (hunk 8 above)
Comment (csessh) [reply snippet/snippet.go:60]: Must be an error on my part while resolving conflicts. I'll have it removed

File: snippet/snippet.go (hunk 7 above)
Comment (csessh) [reply snippet/snippet.go:36]: Good point.
They are also expanded in line 69 below: https://github.com/csessh/pet/blob/5b28fbbb1961049d744b4e7ccd9d94e655e4b91c/snippet/snippet.go#L69

File: config/config.go (hunk 5 above)
Comment (csessh) [reply config/config.go:189]: fair point. 

perhaps keeping the original conditions wouldn't be too bad:
`if s[0] == '~' && os.IsPathSeparator(s[1])`

File: config/config.go (hunk 4 above)
Comment (csessh) [reply config/config.go:103]: If I recall correctly, that's because it would lead to new snippet filename being incorrect. 
https://github.com/knqyf263/pet/blob/adfb69a4f874f4d9027176282cc9482eec8afc05/snippet/snippet.go#L94

for example: 

```
/var/tmp/petNewCommand.toml
```

instead of 

```
/var/tmp/pet/NewCommand.toml
```

File: cmd/new.go [hunk 9]
```diff
@@ -166,12 +166,8 @@ func createAndEditSnippet(newSnippet snippet.SnippetInfo, snippets snippet.Snipp
...
 
 func countSnippetLines() int {
-	filepath, err := config.ExpandPath(config.Conf.General.SnippetFile)
-	if err != nil {
-		return 0
-	}
-
-	f, err := os.Open(filepath)
+	// Count lines in snippet file
+	f, err := os.Open(config.Conf.General.SnippetFile)
```
Comment (RamiAwar) [inline cmd/new.go:170]: How come we don't need the expand here anymore?

Comment (RamiAwar) [reply cmd/new.go:170]: I think we need to find a way to document the expand usage/lack of usage - it's tricky to understand. 

I feel I've asked this same question or the same question reversed quite a few times, and I'm someone familiar with this PR and codebase 😄 

A comment is obviously easiest, but wonder if there's a better way even.

File: config/config.go [hunk 10]
```diff
@@ -186,13 +186,14 @@ func ExpandPath(path string) (string, error) {
 		return path, error
 	}
 
-	if strings.HasPrefix(path, "~/") {
+	if strings.HasPrefix(path, "~") && os.IsPathSeparator(path[1]) {
```
Comment (RamiAwar) [inline config/config.go:189]: Nice one. Also, people who named their files starting with "~" can go to hell 😂

File: snippet/snippet.go (hunk 7 above)
Comment (RamiAwar) [reply snippet/snippet.go]: Ah I see. Hmm. There's a lot of expansions everywhere, sometimes maybe double expansions also. We should check that expansion only happens at the point where it's necessary, i.e. maybe only at line 69 here.

Are the expansions before that line necessary?

File: config/config.go (hunk 4 above)
Comment (RamiAwar) [reply config/config.go:103]: Hmm.. I would say this is a problem on the side of consuming the `snippetdirs` then. When consuming a directory we should make sure to join with os-specific file separators.

I think this is unnecessary if that part works well.

File: snippet/snippet.go [hunk 11]
```diff
@@ -30,10 +30,8 @@ func (snippets *Snippets) Load(includeDirs bool) error {
 	// Create a list of snippet files to load snippets from
 	var snippetFiles []string
 
-	// Load snippets from the main snippet file
-	// Raise an error if the file is not found / not configured
-	snippetFile := config.Conf.General.SnippetFile
-	if snippetFile != "" {
+	snippetFile, err := config.ExpandPath(config.Conf.General.SnippetFile)
```
Comment (RamiAwar) [inline snippet/snippet.go:33]: Keep the first comment please 😄 nice to have even though obvious to us

File: snippet/snippet.go [hunk 12]
```diff
@@ -43,32 +41,41 @@ func (snippets *Snippets) Load(includeDirs bool) error {
...
 
-	// Optionally load snippets from snippet directories
-	if includeDirs {
-		for _, dir := range config.Conf.General.SnippetDirs {
-			if _, err := os.Stat(dir); err != nil {
-				if os.IsNotExist(err) {
-					return fmt.Errorf("snippet directory not found. %s", dir)
-				}
-				return fmt.Errorf("failed to load snippet directory. %v", err)
+	for _, snippetDir := range config.Conf.General.SnippetDirs {
```
Comment (RamiAwar) [inline snippet/snippet.go:49]: I think you removed the includeDirs if statement by mistake. That's actually very important - we don't always want to load all the snippets.

Specifically, when creating a new snippet, we want to load the snippets from the main file only, count the lines, and then add the new snippet to that subset of snippets (main file snippets only) and then save them all into the main file.

If we didn't have includeDirs we would actually save the snippets from ALL dirs into the main file 😓  uncovered this bug recently.

**Review by RamiAwar — CHANGES_REQUESTED** [review]
I wonder if we should wrap all the os.ReadFile, os.WriteFile, etc. operations with our own wrappers that include ExpandPath.

If we add expand path in every single place, then it would make sense. Otherwise no.

File: config/config.go (hunk 5 above)
Comment (RamiAwar) [reply config/config.go]: I would even take this a bit further: turn it into a small function with a nice name. 

IsRelativePath(string)

Then we can write some tests for it and make sure it works as expected, and it can be re-used elsewhere too. But mainly, it is more readable. When I read `isRelativePath(path)` i directly understand what's going on.

vs reading `if s[0] == '~' && os.IsPathSeparator(s[1])` needs at least a tiny bit of thinking more.

less thinking == better code

File: cmd/new.go (hunk 9 above)
Comment (csessh) [reply cmd/new.go:170]: Yeah, we've circled back to this a couple of times. it definitely warrants some better way to document it. 
I'll see what I can come up with.

File: config/config.go (hunk 4 above)
Comment (csessh) [reply config/config.go:103]: I can leave it out for a different PR if that helps

File: snippet/snippet.go (hunk 12 above)
Comment (csessh) [reply snippet/snippet.go:49]: You're right. I definitely did that. Easy enough a fix

File: cmd/new.go (hunk 9 above)
Comment (csessh) [reply cmd/new.go:170]: hate to admit it, but it may have been a bad rebase job on my end. 
Expanding path when reading/accessing file is what we want.

Comment (RamiAwar) [reply cmd/new.go:170]: Yeah I thought so, I'm reviewing this thoroughly though don't worry 🙂

Comment (csessh) [reply cmd/new.go]: so, I tried wrapping those os calls, but they pose a problem, which I'm not particularly fond of: our wrapper functions would hide away the original documentation (think mouse hover over function).

I made sure we add Expand() directly where accessing a file / a dir is required.

Comment (RamiAwar) [reply cmd/new.go]: My only concern with our current approach is this:

Expansion is always necessary when we use the SnippetFile property in our config. Always.
So right now we are leaving room for devs to forget to call Expand.

It would be really nice if we find a way to automatically expand this filename wherever it is accessed. Think "getter method" on `config.Conf.General.SnippetFile`. Maybe an interface method on GeneralConfig - GetSnippetFilePath(). What do you think of that approach? Any other ideas to deal with this?

File: cmd/edit.go [hunk 13]
```diff
@@ -64,7 +64,7 @@ func edit(cmd *cobra.Command, args []string) (err error) {
 }
 
 func fileContent(fname string) string {
-	data, _ := os.ReadFile(fname)
+	data, _ := os.ReadFile(config.Expand(fname))
```
Comment (RamiAwar) [inline cmd/edit.go:67]: It's not necessary to put this here actually - we should just make sure that all the places that call fileContent pass in the expanded path.

File: config/config.go [hunk 14]
```diff
@@ -86,9 +86,9 @@ type FlagConfig struct {
 
 // Load loads a config toml
 func (cfg *Config) Load(file string) error {
-	_, err := os.Stat(file)
+	_, err := os.Stat(Expand(file))
```
Comment (RamiAwar) [inline config/config.go:89]: It's also not necessary to call Expand here and in the readfile below I think.

Look at the code, config.Load is only called in one place, and there it is set to the default location, unless someone passes a command line argument.

I think it is fair to assume that the command line argument will never be a relative path.

Comment (csessh) [reply config/config.go:89]: It's better to be explicit and consistent across all instances, I think. Otherwise, we'd be circling back to "why not expand here" down the road .

Plus expand isn't computationally expensive in any way

File: cmd/new.go (hunk 9 above)
Comment (csessh) [reply cmd/new.go]: There are two use cases for SnippetFile: full path when a file is to be accessed and relative path in any other cases. Problem with a single getter is that it is still tricky to indicate which version to return. 

We can do two getters, or simpler with an additional field called `SnippetFileAbsPath`. This field is hidden from user, auto populated from `SnippetFile` value when config is loaded and it is used specifically for accessing file (documentation could help)

thoughts?

File: config/config.go (hunk 14 above)
Comment (RamiAwar) [reply config/config.go:89]: Totally agree with being explicit and consistent. I'm just against calling it **unnecessarily**, which is a different thing than if it were **optional**.

If we're expanding where we shouldn't expand, it's bad logic. Confusing to the reader, potentially cause for bugs in the future, etc.

Let me think about this some more. There must be a better way. <insert Raymond Hettinger meme>

Comment (RamiAwar) [reply config/config.go:89]: I'm gonna do a few modifications to this, pushing in bit. A nice compromise I found is just using a variable here - makes it clear that the path to be used is the absolute. So anyone modifying this in the future will reuse absPath basically, instead of having to remember to call Expand. This reduces the risk enough to forget about this issue in my eyes 🙂.

Sorry hard to communicate these things exactly - it's more an art trying to balance between preventing future developers from making preventable mistakes, readability, complexity and performance.

**Comment by RamiAwar** [comment]
@csessh Refactored the paths into an AbsolutePath interface and then used that in all places where a path is needed - this way it's always guaranteed to be an abs path when you use it, and it's less dependent on the developers calling Expand() in the right places.

The one downside that is still there is that we still have to convert the config file paths ourselves by constructing this struct, but that can be improved upon later as well.

I also realized we don't need IsRelative because we have IsAbsolute in the standard library, we could just do NOT that. Less custom but probably for the best.

Haven't tested this thoroughly yet - what do you think of the direction though?

**Comment by csessh** [comment]
@RamiAwar I've been a little preoccupied with other things and I haven't got a chance to catch up with your latest messages yet. I'll circle back tomorrow. Thanks for taking the time to refactor, I'll check it out.

**Review by RamiAwar — APPROVED** [review]

**Comment by csessh** [comment]
@RamiAwar This is awesome and I'm onboard with it. (although I feel bad that you had to take over the task)

As far as testing goes:

- I've run the test cases (listed previously) on Linux and it works.
- I got this when testing on Windows: 
```
csessh-pc :: ~/pet ‹feat-syspath› » go run main.go configure                                                                                                                            
# github.com/knqyf263/pet/cmd
cmd\configure.go:23:26: cannot use configFilePath (variable of type "github.com/knqyf263/pet/path".AbsolutePath) as string value in argument to editFile
cmd\edit.go:50:25: cannot use snippetFilePath (variable of type "github.com/knqyf263/pet/path".AbsolutePath) as string value in argument to editFile
cmd\new.go:165:25: cannot use snippetFilePath (variable of type "github.com/knqyf263/pet/path".AbsolutePath) as string value in argument to **editFile**
```

**Comment by RamiAwar** [comment]
Noo don't feel bad at all! It was your amazing PR that got us here in the first place, I didn't take over at all 🙂

Ah nice you ran it on windows, I was still planning on doing that before merging. I think a .Get() is missing still or something.

**Comment by RamiAwar** [comment]
> @RamiAwar This is awesome and I'm onboard with it. (although I feel bad that you had to take over the task, but I'm also glad this PR gets to where it needs to be)
> 
> As far as testing goes:
> 
> * I've run the test cases (listed previously) on Linux and it works.
> * I got this when testing on Windows:
> 
> ```
> csessh-pc :: ~/pet ‹feat-syspath› » go run main.go configure                                                                                                                            
> # github.com/knqyf263/pet/cmd
> cmd\configure.go:23:26: cannot use configFilePath (variable of type "github.com/knqyf263/pet/path".AbsolutePath) as string value in argument to editFile
> cmd\edit.go:50:25: cannot use snippetFilePath (variable of type "github.com/knqyf263/pet/path".AbsolutePath) as string value in argument to editFile
> cmd\new.go:165:25: cannot use snippetFilePath (variable of type "github.com/knqyf263/pet/path".AbsolutePath) as string value in argument to **editFile**
> ```

Managed to fix that windows editFile error

**Comment by csessh** [comment]
> > @RamiAwar This is awesome and I'm onboard with it. (although I feel bad that you had to take over the task, but I'm also glad this PR gets to where it needs to be)
> > As far as testing goes:
> > 
> > * I've run the test cases (listed previously) on Linux and it works.
> > * I got this when testing on Windows:
> > 
> > ```
> > csessh-pc :: ~/pet ‹feat-syspath› » go run main.go configure                                                                                                                            
> > # github.com/knqyf263/pet/cmd
> > cmd\configure.go:23:26: cannot use configFilePath (variable of type "github.com/knqyf263/pet/path".AbsolutePath) as string value in argument to editFile
> > cmd\edit.go:50:25: cannot use snippetFilePath (variable of type "github.com/knqyf263/pet/path".AbsolutePath) as string value in argument to editFile
> > cmd\new.go:165:25: cannot use snippetFilePath (variable of type "github.com/knqyf263/pet/path".AbsolutePath) as string value in argument to **editFile**
> > ```
> 
> Managed to fix that windows editFile error

I can confirm.
I'll run those tests again on Windows

**Comment by RamiAwar** [comment]
Your comments made me question why we don't have a test pipeline for Windows 😄 was a one-liner to add.

Also turned out other tests were failing. Looks OK now!

**Comment by npearson72** [comment]
I can't say for sure, but I think this broke the ability to pass env vars into the `snippetfile`

My config is: `snippetfile = "$XDG_CONFIG_HOME/pet/snippet.toml"`

`$XDG_CONFIG_HOME` this points to an absolute directory.

This worked in the past. I haven't updated in a while, but just recently did and noticed I'm getting this error:

```
load snippet failed: snippet file not found. $XDG_CONFIG_HOME/pet/snippet.toml
Please run 'pet configure' and provide a correct file path, or remove this
if you only want to provide snippetdirs instead
```

**Comment by RamiAwar** [comment]
> dated in a while, but just recently did and noticed I'm getting this e

Interesting, will file an issue for testing with this. Our test suite is still relatively new and I don't think we have this usecase covered in a test - we can try adding it and see if it fails.

Thanks for reporting this!
